docs/modules/serialization/pages/schema-evolution-full.adoc

```diff
@@ -0,0 +1,779 @@
+* Always write all defined fields, even if their values are `null` or default.
+* Avoid conditional or state-dependent logic inside the `write` method.
+* Keep field names and their corresponding `FieldKind` stable between releases.
+
+=== Schema evolution approaches
+
+Hazelcast supports two primary approaches to managing schema evolution, depending on operational needs and compatibility requirements.
+
+[cols="1,3", options="header"]
+|===
+| Approach | Description
+
+| **Single IMap Evolution**
+| All schema changes occur within the same IMap and `typeName`. This approach allows members to coexist using different schema versions during rolling upgrades. It is suitable for additive or compatible changes (for example, adding new fields).
+
+| **Versioned Maps**
+| Each schema version is stored in a separate map (for example, `Order_v1`, `Order_v2`). This pattern isolates incompatible changes such as field renames, type narrowing, or semantic changes. Data can then be migrated using a pipeline or map-to-map transformation before decommissioning the older map. Unlike schema evolution handled transparently by Compact serialization, this pattern requires explicit design changes in the user application, for example, managing multiple map names and handling data migration logic, but provides greater control when compatibility cannot be automatically maintained.
+
+|===
+
+Choosing between these two strategies depends on the degree of backward compatibility required.
+In general:
+
+* Use **Single Map Evolution** for additive, backward-compatible changes.
+* Use **Versioned Maps** for incompatible or semantically breaking changes.
+
+=== Compact schema propagation
+
+Each Compact schema version is identified by its fingerprint and propagated automatically across the cluster when new data is written.
+Members and clients cache these schemas locally and can fetch missing versions on demand.
+This mechanism allows heterogeneous environments, where not all members are upgraded simultaneously, to continue operating safely during rolling upgrades or failover scenarios.
+
+=== Consistency and partitioning
+
+When Compact-serialized objects are used as map keys, any change to the fields that participate in partitioning (for example, the key field or composite key fields) alters the data distribution across the cluster.
+Such changes are considered **incompatible** and require migration or rekeying.
+Hazelcast treats these as structural rather than schema-level changes, and they must be handled through data transformation pipelines or controlled rebalancing processes.
+
+== Decision flow for schema changes
+
+When evolving Compact-serialized objects, different kinds of changes have different effects on compatibility and runtime behavior.
+This section helps determine whether a change is **safe**, **requires caution**, or **is incompatible** and needs migration.
+
+=== Overview
+
+Schema changes fall into five broad categories:
+
+1. Adding or removing a field
```

Comment:
there is a significant difference between adding mandatory and optional field. you described only optional field